File: wrappers/php/src/ChatTemplate.php
<?php

declare(strict_types=1);

namespace TurboToken;

class ChatMessage
{
    public string $role;
    public string $content;
    public ?string $name;

    public function __construct(string $role, string $content, ?string $name = null)
    {
        $this->role = $role;
        $this->content = $content;
        $this->name = $name;
    }
}

class ChatTemplate
{
    public string $messagePrefix;
    public string $messageSuffix;
    public string $assistantPrefix;

    public function __construct(
        string $messagePrefix,
        string $messageSuffix,
        string $assistantPrefix
    ) {
        $this->messagePrefix = $messagePrefix;
        $this->messageSuffix = $messageSuffix;
        $this->assistantPrefix = $assistantPrefix;
    }

    /**
     * @param ChatMessage[] $messages
     */
    public function formatMessages(array $messages): string
    {
        $text = '';
        foreach ($messages as $msg) {
            $nameTag = $msg->name !== null ? " name={$msg->name}" : '';
            $text .= $this->messagePrefix
                . "<|im_start|>{$msg->role}{$nameTag}\n"
                . $msg->content
                . "<|im_end|>"
                . $this->messageSuffix;
        }
        $text .= $this->assistantPrefix;
        return $text;
    }

    public static function resolveTurboTokenV1(): self
    {
        return new self('', "\n", "<|im_start|>assistant\n");
    }

    public static function resolveImTokens(): self
    {
        return new self('', "\n", "<|im_start|>assistant\n");
    }
}

File: wrappers/kotlin/src/main/kotlin/com/turbotoken/ChatTemplate.kt
@file:JvmName("KChatTemplate")
package com.turbotoken

/**
 * Kotlin data classes for chat message formatting.
 */
data class KChatMessage(
    val role: String,
    val content: String,
    val name: String? = null
) {
    /** Convert to the Java ChatMessage for interop. */
    fun toJava(): ChatTemplate.ChatMessage =
        ChatTemplate.ChatMessage(role, name, content)
}

data class KChatTemplate(
    val messagePrefix: String,
    val messageSuffix: String,
    val assistantPrefix: String
)

/**
 * Sealed class for template mode selection.
 */
sealed class KTemplateMode {
    object TurbotokenV1 : KTemplateMode()
    object ImTokens : KTemplateMode()

    fun toJava(): ChatTemplate.TemplateMode = when (this) {
        is TurbotokenV1 -> ChatTemplate.TemplateMode.TURBOTOKEN_V1
        is ImTokens -> ChatTemplate.TemplateMode.IM_TOKENS
    }
}

data class KChatOptions(
    val primeWithAssistantResponse: Boolean = false,
    val templateMode: KTemplateMode = KTemplateMode.TurbotokenV1
) {
    fun toJava(): ChatTemplate.ChatOptions =
        ChatTemplate.ChatOptions(primeWithAssistantResponse, templateMode.toJava())
}

/**
 * Resolves the chat template for the given mode.
 */
fun resolveTemplate(mode: KTemplateMode): KChatTemplate {
    val java = ChatTemplate.resolve(mode.toJava())
    return KChatTemplate(
        messagePrefix = java.messagePrefix,
        messageSuffix = java.messageSuffix,
        assistantPrefix = java.assistantPrefix
    )
}

/**
 * Formats chat messages into a string for tokenization.
 */
fun formatMessages(messages: List<KChatMessage>, options: KChatOptions = KChatOptions()): String {
    val javaMessages = messages.map { it.toJava() }
    return ChatTemplate.formatMessages(javaMessages, options.toJava())
}
